README.md
<span align="left"><img src="https://github.com/luc-github/ESP3D/blob/2.1/images/ESP3D.png" width="200px"/></span><span align="left">Firmware for ESP8266/ESP8285  and ESP32 used with 3D printer</span>    
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[<img src="https://img.shields.io/liberapay/patrons/ESP3D.svg?logo=liberapay">](https://liberapay.com/ESP3D)

[Latest stable release ![Release Version](https://img.shields.io/github/release/luc-github/ESP3D.svg?style=plastic) ![Release Date](https://img.shields.io/github/release-date/luc-github/ESP3D.svg?style=plastic)](https://github.com/luc-github/ESP3D/releases/latest/) [![github-ci](https://github.com/luc-github/ESP3D/workflows/build-ci-2.0/badge.svg)](https://github.com/luc-github/ESP3D/actions/workflows/build-ci-2.0.yml) [![Release Version](https://img.shields.io/github/v/release/luc-github/ESP3D-WEBUI?color=green&include_prereleases&label=WebUI&style=plastic)](https://github.com/luc-github/ESP3D-WEBUI/tree/2.1)      
please use Arduino ide 1.8.9+ with [![Release Version](https://img.shields.io/badge/ESP32-1.0.4-green?style=plastic)](https://github.com/espressif/arduino-esp32/releases/tag/1.0.4) or [![Release Version](https://img.shields.io/badge/ESP8266-2.5.2-green?style=plastic)](https://github.com/esp8266/Arduino/releases/tag/2.5.2)

[Latest development version ![Development Version](https://img.shields.io/badge/Devt-v3.0-yellow?style=plastic) ![GitHub last commit (branch)](https://img.shields.io/github/last-commit/luc-github/ESP3D/3.0?style=plastic)](https://github.com/luc-github/ESP3D/tree/3.0) [![github-ci](https://github.com/luc-github/ESP3D/workflows/build-ci/badge.svg)](https://github.com/luc-github/ESP3D/actions/workflows/build-ci.yml) [![Development  Version](https://img.shields.io/badge/Devt-v3.0-yellow?style=plastic&label=WebUI)](https://github.com/luc-github/ESP3D-WEBUI/tree/3.0)   
please use Arduino ide 1.8.9+ with [![Release Version](https://img.shields.io/badge/ESP32-git-yellow?style=plastic&logo=github)](https://github.com/espressif/arduino-esp32) or [![Release Version](https://img.shields.io/badge/ESP8266-git-yellow?style=plastic&logo=github)](https://github.com/esp8266/Arduino/)

[Project board](https://github.com/users/luc-github/projects/1/views/1)

[All releases](https://github.com/luc-github/ESP3D/releases)

This firmware allows not only to have a cheap bridge between Wifi and serial, but also to have a web UI to configure wifi, to monitor 3D printer and even control it, and to make things easy,
UI is fully customizable without reflashing FW.

Firmware should work with any 3D printer firmware (repetier/marlin/smoothieware using GCODE) if serial connection has a correct setup.
I currently use it with my personnal flavor of [repetier for Due based boards](https://github.com/luc-github/Repetier-Firmware-0.92).

The web interface files are present in data directory but UI has it's own repository [ESP3D-WEBUI](https://github.com/luc-github/ESP3D-WEBUI).
* Be aware  ESP3D-WEBUI is for firmware 0.9.99 minimum - previous released version use tpl files which are no more used.
* Note for ESP8266 1MB flash : FW is now too big will all features you need to chose strip the FW and select only some features, also WebUI is now also too big for full multilanguage support to fit the 128K SPIFFS so please use pack with limited language (en +another) https://github.com/luc-github/ESP3D-WEBUI/tree/2.1/languages
    

## Sponsors 
[<img width="200px" src="https://raw.githubusercontent.com/luc-github/ESP3D-WEBUI/2.1/images/sponsors-supporters/MKS/mksmakerbase.jpg" title="MKS Makerbase">](https://github.com/makerbase-mks)&nbsp;&nbsp;

## Supporters

## Become a sponsor or a supporter
 * A sponsor is a recurent donator    
If your tier is `10 US$/month` or more, to thank you for your support, your logo / avatar will be added to the readme page with eventually with a link to your site.    
 * A supporter is per time donator 
 If your donation is over `120 US$` per year, to thank you for your support, your logo / avatar will be added to the readme page with eventually with a link to your site.  

 Every support is welcome, indeed helping users / developing new features need time and devices, donations contribute a lot to make things happen, thank you.

* liberapay <a href="https://liberapay.com/ESP3D/donate"><img alt="Donate using Liberapay" src="https://liberapay.com/assets/widgets/donate.svg"></a> 
* Paypal [<img src="https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG_global.gif" border="0" alt="PayPal – The safer, easier way to pay online.">](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=FQL59C749A78L)
* ko-fi [![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/G2G0C0QT7)   

## Features
* Serial/Wifi bridge using configurable port 8888
* Use GPIO2 to ground to reset all settings in hard way - 2-6 sec after boot / not before!! Set GPIO2 to ground before boot change boot mode and go to special boot that do not reach FW. Currently boot take 10 sec - giving 8 seconds to connect GPIO2 to GND and do a hard recovery for settings
* Complete configuration by web browser (Station or Access point) or by Serial/telnet commands
* Authentication (optional) for better security
* Update firmware by web browser
* Captive portal in Access point mode which redirect all unknow call to main page
* mDNS which allows to key the name defined in web browser and connect only with bonjour installed on computer
* SSDP, this feature is a discovery protocol, supported on Windows out of the box
* Fail safe mode (Access point)is enabled if cannot connect to defined station at boot.
* Embedded FS uploader and updater.
* OLED screen support  
* Notifications using Line / Pushover / email   
* The web ui add even more feature : https://github.com/luc-github/ESP3D-WEBUI/blob/2.1/README.md#features  

## Default Configuration      
Default Settings:    
* Access Point: ESP3D    
* PW:12345678   
* Authentification: WPA     
* IP: 192.168.0.1   
* Baud rate: 115200   
* Web port:80 
* Data port: 8888     
if Authentication is enabled :
* User: admin   
* Password: admin   
* User:user   
* Password: user   



## Direct commands:    
Check wiki : https://github.com/luc-github/ESP3D/wiki/Direct-ESP3D-commands

## Installation instructions
See [Wiki](https://github.com/luc-github/ESP3D/wiki/Install-Instructions)

## Installation
Feedback on 2.0 was : ESP3D being a library is not really useful and make setup more complex, so now we are back to simple application.   

1. If you haven't already set up Arduino IDE for ESPs then do so for [ESP8266](https://github.com/esp8266/Arduino) or [ESP32 core version](https://github.com/espressif/arduino-esp32).
2. Download the [latest release](https://github.com/luc-github/ESP3D/releases/) and manually copy the libraries present in the `ESP3D-x.y.z/libraries` directory into your `Arduino/libraries` directory.  (*no need if using platformIO*).  *These versions are verified to work with ESP3D, any others (newer version) may cause untested behavior.*
- Use webserver support (recommended as stable), asyncwebserver support is no longer stable on ESP3D 
- * arduinoWebSockets from @Links2004
 
Generic ones:      
Specific for ESP32    
* ESP32SSDP
If you want OLED support:  
* oled-ssd1306 from @squix78    

If you want DHT11/22 support:  
* DHT_sensor_library_for_ESPx from @beegee-tokyo   
3. Compile project esp3d.ino according target: ESP8266 board or ESP32 board, please review config.h to enable disable a feature, by default athentication is disabled and most others are enabled.   
* for ESP8266 set CPU freq to 160MHz for better (https://github.com/luc-github/ESP3D/wiki/Install-Instructions)
4. Upload the data content on ESP3D file system
* Using SPIFFS uploader, this plugin and install instructions is available on each ESP core - please refere to it
or
* Using embedded uploader (you may need to format SPIFFS using : [ESP710]FORMAT on ESP8266 first)    
if embedded uploader does not show up you can force it ti display using : http://your_IP_address?forcefallback=yes    
<img src=https://raw.githubusercontent.com/luc-github/ESP3D/2.1/images/docs/embedded.png><br>


## Update
* Generate a binary using the export binary menu from Arduino IDE and upload it using ESP-WEBUI or embedded interface  

<H3>:warning:Do not flash your Printer fw with ESP connected on Serial - it bring troubles, at least on DaVinci, but no issue if you update using web UI</H3>

## Contribution/customization
* To style the code before pushing PR please use [astyle --style=otbs *.h *.cpp *.ino](http://astyle.sourceforge.net/)   
* The embedded page is created using nodejs then gulp to generate a compressed html page (tool.html.gz), all necessary modules will be installed using the build.bat, you also need bin2c tool (https://sourceforge.net/projects/bin2c/) to generate the h file from the binary,  installation and build is done using the build.bat.   
* The corresponding UI is located [here](https://github.com/luc-github/ESP3D-WEBUI/tree/2.1)
* An optional UI was development using old repetier UI - check [UI\repetier\testui.htm] (https://raw.githubusercontent.com/wiki/luc-github/ESP3D/UI/repetier/testui.htm) file   

 
## Need more information about supported boards or wiring ?
[Check the wiki](https://github.com/luc-github/ESP3D/wiki)

## :question:Any question ?   
Check [Wiki](https://github.com/luc-github/ESP3D/wiki/Install-Instructions) or Join the chat at [![Discord server](https://img.shields.io/discord/752822148795596940?color=blue&label=discord&logo=discord)](https://discord.gg/Z4ujTwE)   

## :exclamation:Any issue/feedback ?    
Check [Wiki](https://github.com/luc-github/ESP3D/wiki/Install-Instructions) and [FAQ](https://github.com/luc-github/ESP3D/discussions?discussions_q=category%3AF.A.Q) 

Changelog.txt
## What's Changed

# Updates
* Update CI to use github-actions instead of Travis
* Update platformio.ini to support more features
* Create wiki source directory and github-action to sync with wiki
* Add ESP01S in platformIO thanks @padcom
* Added support for IFTTT Webhook integration thanks @KingBain

# Fixes
* Fix compilation issues with esp32 core 1.0.5
* Fix wrong detection of [ESPXXX]Command if used in macro or web command and sent to printer using M118
* Fix warnings in updater only mode
* Fix several warnings in code and libs 
* Fix typo in pushError function (wrong type)

# Wiki / Documentation / Readme
* Add connection diagram for weedo tina2 thanks @atsju
* Add recommended extensions for easier linting thanks @atsju
* Update feature request template thanks @atsju
* Update Wiki / Readme thanks @atsju
* Update Readme thanks @jschwalbe
* Update funding methods
* Update chat from gitter to discord
* Update index.html.gz (use english mono language)

**Full Changelog**: https://github.com/luc-github/ESP3D/compare/v2.1.1...v2.1.2

src/config.h
/*
  config.h - ESP3D configuration class

  Copyright (c) 2014 Luc Lebosse. All rights reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

//version and sources location
#define FW_VERSION "2.0.0.c25"
#define REPOSITORY "https://github.com/luc-github/ESP3D"

//Customize ESP3D ////////////////////////////////////////////////////////////////////////
#define ESP8266_MODEL_NAME "ESP8266"
#define ESP8266_MODEL_URL "http://espressif.com/en/products/esp8266/"
#define ESP32_MODEL_NAME "ESP32"
#define ESP32_MODEL_URL "https://www.espressif.com/en/products/hardware/esp-wroom-32/overview"
#define ESP_MODEL_NUMBER "ESP3D 2.0" 
#define ESP_MANUFACTURER_NAME "Espressif Systems"
#define ESP_MANUFACTURER_URL "http://espressif.com"
//default name if no mac address is valid
#define ESP_DEFAULT_NAME "MYESP"
//if commented name will follow mac address 3 last digits
//like ESP_XXXXXX (eg:ESP_028E41) to avoid overlap if several ESP3D
#define ESP_HOST_NAME ESP_DEFAULT_NAME


//FEATURES - comment to disable //////////////////////////////////////////////////////////

//Do we use async webserver or not
//#define ASYNCWEBSERVER

//SERIAL_COMMAND_FEATURE: allow to send command by serial
#define SERIAL_COMMAND_FEATURE

//TCP_IP_DATA_FEATURE: allow to connect serial from TCP/IP
#define TCP_IP_DATA_FEATURE

//MDNS_FEATURE: this feature allow  type the name defined
//in web browser by default: http:\\esp8266.local and connect
#define MDNS_FEATURE

//SSDD_FEATURE: this feature is a discovery protocol, supported on Windows out of the box
#define SSDP_FEATURE

//NETBIOS_FEATURE: this feature is a discovery protocol, supported on Windows out of the box
#define NETBIOS_FEATURE

//CAPTIVE_PORTAL_FEATURE: In SoftAP redirect all unknow call to main page
#define CAPTIVE_PORTAL_FEATURE

//WEB_UPDATE_FEATURE: allow to flash fw using web UI
#define WEB_UPDATE_FEATURE

//RECOVERY_FEATURE: allow to use GPIO2 pin as hardware reset for EEPROM, add 8s to boot time to let user to jump GPIO2 to GND
//#define RECOVERY_FEATURE

//DIRECT_PIN_FEATURE: allow to access pin using ESP201 command
#define DIRECT_PIN_FEATURE

//ESP_OLED_FEATURE: allow oled screen output
//#define ESP_OLED_FEATURE

//DHT_FEATURE: send update of temperature / humidity based on DHT 11/22
//#define DHT_FEATURE

//AUTHENTICATION_FEATURE: protect pages by login password
//#define AUTHENTICATION_FEATURE

//WS_DATA_FEATURE: allow to connect serial from Websocket
#define WS_DATA_FEATURE

//TIMESTAMP_FEATURE: Time stamp feature on direct SD  files
//#define TIMESTAMP_FEATURE

//Extra features /////////////////////////////////////////////////////////////////////////

//Serial rx buffer size is 256 but can be extended
#define SERIAL_RX_BUFFER_SIZE 512

//Serial Parameters
#define ESP_SERIAL_PARAM SERIAL_8N1

//which serial ESP use to communicate to printer (ESP32 has 3 serials available, ESP8266 only one)
//Uncomment one only
#define USE_SERIAL_0
//For ESP32 Only
//#define USE_SERIAL_1
//#define USE_SERIAL_2

//Pins Definition ////////////////////////////////////////////////////////////////////////
//-1 means use default pins of your board what ever the serial you choose
#define ESP_RX_PIN -1
#define ESP_TX_PIN -1

#ifdef RECOVERY_FEATURE
//pin used to reset setting
#define RESET_CONFIG_PIN 2
#endif

#ifdef DHT_FEATURE
#define ESP_DHT_PIN 15
#endif

//Pins where the screen is connected
#ifdef ESP_OLED_FEATURE
#define OLED_DISPLAY_SSD1306  // OLED Display Type: SSD1306(OLED_DISPLAY_SSD1306) / SH1106(OLED_DISPLAY_SH1106), comment this line out to disable oled
#define OLED_PIN_SDA  4  //5 //SDA;  // i2c SDA Pin
#define OLED_PIN_SCL  15  //4 //SCL;  // i2c SCL Pin
#define OLED_ADDR	0x3c
#define HELTEC_EMBEDDED_PIN 16 //0 to disable
#define OLED_FLIP_VERTICALY 1 //0 to disable
#endif


//Supported FW /////////////////////////////////////////////////////////////
#define UNKNOWN_FW 0
#define REPETIER4DV	1
#define MARLIN		2
#define MARLINKIMBRA		3
#define SMOOTHIEWARE	4
#define REPETIER		5
#define GRBL		6
#define MAX_FW_ID 6

//Do not Edit after this line //////////////////////////////////////////////

//DEBUG Flag do not do this when connected to printer !!!
//be noted all upload may failed if enabled
//#define DEBUG_ESP3D
//#define DEBUG_OUTPUT_SPIFFS
//#define DEBUG_OUTPUT_SERIAL
//#define DEBUG_OUTPUT_TCP

//Sanity check
#ifndef SDCARD_FEATURE
#ifdef TIMESTAMP_FEATURE
#undef TIMESTAMP_FEATURE
#endif
#endif

#if defined(ASYNCWEBSERVER)
#define ESP_USE_ASYNC true
#else
#define ESP_USE_ASYNC false
#endif

//number of clients allowed to use data port at once
#define MAX_SRV_CLIENTS 1

#ifdef ARDUINO_ARCH_ESP32
#include "FS.h"
#include "SPIFFS.h"
using fs::File;
#define WIFI_NONE_SLEEP WIFI_PS_NONE
#define WIFI_LIGHT_SLEEP WIFI_PS_MIN_MODEM
#define WIFI_MODEM_SLEEP WIFI_PS_MAX_MODEM
#define WIFI_PHY_MODE_11B WIFI_PROTOCOL_11B
#define WIFI_PHY_MODE_11G WIFI_PROTOCOL_11B | WIFI_PROTOCOL_11G
#define WIFI_PHY_MODE_11N WIFI_PROTOCOL_11B | WIFI_PROTOCOL_11G | WIFI_PROTOCOL_11N
#define AUTH_OPEN WIFI_AUTH_OPEN
#define AUTH_WEP WIFI_AUTH_WEP
#define AUTH_WPA_PSK WIFI_AUTH_WPA_PSK
#define AUTH_WPA2_PSK WIFI_AUTH_WPA2_PSK
#define AUTH_WPA_WPA2_PSK WIFI_AUTH_WPA_WPA2_PSK
#define ENC_TYPE_NONE AUTH_OPEN
#define FS_FILE File
#define FS_DIR File
#define SD_FILE_READ FILE_READ
#define SPIFFS_FILE_READ FILE_READ
#define SD_FILE_WRITE FILE_WRITE
#define SPIFFS_FILE_WRITE FILE_WRITE
#define WIFI_EVENT_STAMODE_CONNECTED SYSTEM_EVENT_STA_CONNECTED
#define WIFI_EVENT_STAMODE_DISCONNECTED SYSTEM_EVENT_STA_DISCONNECTED
#define WIFI_EVENT_STAMODE_GOT_IP SYSTEM_EVENT_STA_GOT_IP
#define WIFI_EVENT_SOFTAPMODE_STACONNECTED SYSTEM_EVENT_AP_STACONNECTED
#else
#define FS_DIR fs::Dir
#define FS_FILE fs::File
#define SD_FILE_READ FILE_READ
#define SPIFFS_FILE_READ "r"
#define SD_FILE_WRITE FILE_WRITE
#define SPIFFS_FILE_WRITE "w"
#endif



//WEBHOST_SDCARD_FEATURE : to use SDCard to host webpages
//NOT YET IMPLEMENTED!!! Keep it as TODO
//#define WEBHOST_SDCARD_FEATURE

#ifdef DEBUG_ESP3D
#ifdef DEBUG_OUTPUT_SPIFFS
#ifndef FS_NO_GLOBALS
#define FS_NO_GLOBALS
#endif
#include <FS.h>
#define DEBUG_PIPE NO_PIPE
#define LOG(string) { FS_FILE logfile = SPIFFS.open("/log.txt", "a+");logfile.print(string);logfile.close();}
#endif

#ifdef DEBUG_OUTPUT_SERIAL
#define DEBUG_PIPE SERIAL_PIPE
#define LOG(string) {Serial.print(string);}
#endif
#ifdef DEBUG_OUTPUT_TCP
#include "espcom.h"
#define LOG(string) {ESPCOM::send2TCP(string, false);}
#define DEBUG_PIPE TCP_PIPE
#endif
#else
#define LOG(string) {}
#define DEBUG_PIPE NO_PIPE
#endif

#define NOLOG(string) {}

#ifndef CONFIG_h
#define CONFIG_h

#include <Arduino.h>
#ifdef ARDUINO_ARCH_ESP8266
extern "C" {
#include "user_interface.h"
}
#else

#endif
#include "wificonf.h"

typedef enum {
    UPLOAD_STATUS_NONE = 0,
    UPLOAD_STATUS_FAILED = 1,
    UPLOAD_STATUS_CANCELLED = 2,
    UPLOAD_STATUS_SUCCESSFUL = 3,
    UPLOAD_STATUS_ONGOING  = 4
} upload_status_type;

typedef enum {
    NO_PIPE = 0,
    SERIAL_PIPE = 2,
    SERIAL1_PIPE = 3,
    SERIAL2_PIPE = 4,
#ifdef TCP_IP_DATA_FEATURE
    TCP_PIPE = 5,
#endif
#ifdef WS_DATA_FEATURE
    WS_PIPE = 6,
#endif
#ifdef ESP_OLED_FEATURE
    OLED_PIPE = 7,
#endif
    WEB_PIPE = 8,
    PRINTER_PIPE = 9
} tpipe;

#define DEFAULT_PRINTER_PIPE SERIAL_PIPE

typedef enum {
    LEVEL_GUEST = 0,
    LEVEL_USER = 1,
    LEVEL_ADMIN = 2
} level_authenticate_type;


#define    NO_SD 0
#define    SD_DIRECTORY 1
#define    EXT_DIRECTORY 2


//flags
#define AP_MODE			1
#define CLIENT_MODE		2
#define DHCP_MODE		1
#define STATIC_IP_MODE		2

//position in EEPROM
//AP mode = 1; Station client mode = 2
#define EP_WIFI_MODE			0    //1 byte = flag
#define EP_STA_SSID				1    //33 bytes 32+1 = string  ; warning does not support multibyte char like chinese
#define EP_STA_PASSWORD			34   //65 bytes 64 +1 = string ;warning does not support multibyte char like chinese
#define EP_STA_IP_MODE			99   //1 byte = flag
#define EP_STA_IP_VALUE			100  //4  bytes xxx.xxx.xxx.xxx
#define EP_STA_MASK_VALUE			104  //4  bytes xxx.xxx.xxx.xxx
#define EP_STA_GATEWAY_VALUE			108  //4  bytes xxx.xxx.xxx.xxx
#define EP_BAUD_RATE			112  //4  bytes = int
#define EP_STA_PHY_MODE			116  //1 byte = flag
#define EP_SLEEP_MODE			117  //1 byte = flag
#define EP_CHANNEL			118 //1 byte = flag
#define EP_AUTH_TYPE			119 //1 byte = flag
#define EP_SSID_VISIBLE			120 //1 byte = flag
#define EP_WEB_PORT			121 //4  bytes = int
#define EP_DATA_PORT			125 //4  bytes = int
#define EP_OUTPUT_FLAG			129 //1  bytes = flag
#define EP_HOSTNAME				130//33 bytes 32+1 = string  ; warning does not support multibyte char like chinese
#define EP_DHT_INTERVAL		    164//4  bytes = int
#define EP_FREE_INT2		    168//4  bytes = int
#define EP_FREE_INT3		    172//4  bytes = int
#define EP_ADMIN_PWD		    176//21  bytes 20+1 = string  ; warning does not support multibyte char like chinese
#define EP_USER_PWD		    197//21  bytes 20+1 = string  ; warning does not support multibyte char like chinese
#define EP_AP_SSID				218    //33 bytes 32+1 = string  ; warning does not support multibyte char like chinese
#define EP_AP_PASSWORD			251   //65 bytes 64 +1 = string ;warning does not support multibyte char like chinese
#define EP_AP_IP_VALUE			316  //4  bytes xxx.xxx.xxx.xxx
#define EP_AP_MASK_VALUE			320  //4  bytes xxx.xxx.xxx.xxx
#define EP_AP_GATEWAY_VALUE			324  //4  bytes xxx.xxx.xxx.xxx
#define EP_AP_IP_MODE			329   //1 byte = flag
#define EP_AP_PHY_MODE			330  //1 byte = flag
#define EP_SD_SPEED_DIV			331  //1 byte = flag
#define EP_FREE_STRING1			332  //128 bytes 127+1 = string  ; warning does not support multibyte char like chinese
#define EP_DHT_TYPE		460 //1  bytes = flag
#define EP_TARGET_FW		461 //1  bytes = flag
#define EP_TIMEZONE         462//1  bytes = flag
#define EP_TIME_ISDST       463//1  bytes = flag
#define EP_TIME_SERVER1 464//129 bytes 128+1 = string  ; warning does not support multibyte char like chinese  
#define EP_TIME_SERVER2  593 //129 bytes 128+1 = string  ; warning does not support multibyte char like chinese
#define EP_TIME_SERVER3  722 //129 bytes 128+1 = string  ; warning does not support multibyte char like chinese
#define EP_IS_DIRECT_SD   850//1  bytes = flag
#define EP_PRIMARY_SD   851//1  bytes = flag
#define EP_SECONDARY_SD   852//1  bytes = flag
#define EP_DIRECT_SD_CHECK   853//1  bytes = flag
#define EP_SD_CHECK_UPDATE_AT_BOOT   854//1  bytes = flag

#define LAST_EEPROM_ADDRESS 855
//next available is 855
//space left 1024 - 855 = 169
//extra fre
//#define EP_FREE_INT2		    168//4  bytes = int
//#define EP_FREE_INT3		    172//4  bytes = int
//#define EP_FREE_STRING1			331  //129 bytes 128+1 = string  ; warning does not support multibyte char like chinese

//default values
#define DEFAULT_WIFI_MODE			AP_MODE
const char DEFAULT_AP_SSID []  PROGMEM =		"ESP3D";
const char DEFAULT_AP_PASSWORD [] PROGMEM =	"12345678";
const char DEFAULT_STA_SSID []  PROGMEM =		"ESP3D";
const char DEFAULT_STA_PASSWORD [] PROGMEM =	"12345678";
const byte DEFAULT_STA_IP_MODE  = 				DHCP_MODE;
const byte DEFAULT_AP_IP_MODE = 				STATIC_IP_MODE;
const byte DEFAULT_IP_VALUE[]   =	        {192, 168, 0, 1};
const byte DEFAULT_MASK_VALUE[]  =	        {255, 255, 255, 0};
#define DEFAULT_GATEWAY_VALUE   	        DEFAULT_IP_VALUE
const long DEFAULT_BAUD_RATE =			115200;
#define DEFAULT_PHY_MODE			WIFI_PHY_MODE_11G
#define DEFAULT_SLEEP_MODE			WIFI_MODEM_SLEEP
#define DEFAULT_CHANNEL				11
#define DEFAULT_AUTH_TYPE			AUTH_WPA_PSK
#define DEFAULT_SSID_VISIBLE			1
#define DEFAULT_MAX_CONNECTIONS			4
#define DEFAULT_BEACON_INTERVAL			100
const int DEFAULT_WEB_PORT =			80;
const int DEFAULT_DATA_PORT =			8888;
const char DEFAULT_ADMIN_PWD []  PROGMEM =	"admin";
const char DEFAULT_USER_PWD []  PROGMEM =	"user";
const char DEFAULT_ADMIN_LOGIN []  PROGMEM =	"admin";
const char DEFAULT_USER_LOGIN []  PROGMEM =	"user";
const char DEFAULT_TIME_SERVER1 []  PROGMEM =	"time.nist.gov";
const char DEFAULT_TIME_SERVER2 []  PROGMEM =	"0.pool.ntp.org";
const char DEFAULT_TIME_SERVER3 []  PROGMEM =	"1.pool.ntp.org";
#define DEFAULT_TIME_ZONE			0
#define DEFAULT_TIME_DST			0
#define DEFAULT_PRIMARY_SD  2
#define DEFAULT_SECONDARY_SD 1
#define DEFAULT_DIRECT_SD_CHECK 0
#define DEFAULT_SD_CHECK_UPDATE_AT_BOOT 1
#define DEFAULT_OUTPUT_FLAG 0
#define DEFAULT_DHT_TYPE 255
const int DEFAULT_DHT_INTERVAL =			30;

#define DEFAULT_IS_DIRECT_SD 0

//SD Card reader speed
//possible values are :SPI_FULL_SPEED, SPI_DIV3_SPEED,
//SPI_HALF_SPEED, SPI_DIV6_SPEED, SPI_QUARTER_SPEED,
//SPI_EIGHTH_SPEED, SPI_SIXTEENTH_SPEED
//Decrease if reader give error
#ifdef ARDUINO_ARCH_ESP8266
#define DEFAULT_SDREADER_SPEED SPI_HALF_SPEED
#else
#define DEFAULT_SDREADER_SPEED 4
#endif

const uint16_t Setting[][2] = {
    {EP_WIFI_MODE, LEVEL_ADMIN},//0
    {EP_STA_SSID, LEVEL_ADMIN},//1
    {EP_STA_PASSWORD, LEVEL_ADMIN},//2
    {EP_STA_IP_MODE, LEVEL_ADMIN},//3
    {EP_STA_IP_VALUE, LEVEL_ADMIN},//4
    {EP_STA_MASK_VALUE, LEVEL_ADMIN},//5
    {EP_STA_GATEWAY_VALUE, LEVEL_ADMIN},//6
    {EP_BAUD_RATE, LEVEL_ADMIN},//7
    {EP_STA_PHY_MODE, LEVEL_ADMIN},//8
    {EP_SLEEP_MODE, LEVEL_ADMIN},//9
    {EP_CHANNEL, LEVEL_ADMIN},//10
    {EP_AUTH_TYPE, LEVEL_ADMIN},//11
    {EP_SSID_VISIBLE, LEVEL_ADMIN},//12
    {EP_WEB_PORT, LEVEL_ADMIN},//13
    {EP_DATA_PORT, LEVEL_ADMIN},//14
    {EP_HOSTNAME, LEVEL_ADMIN},//15
    {EP_ADMIN_PWD, LEVEL_ADMIN},//16
    {EP_USER_PWD, LEVEL_USER},//17
    {EP_AP_SSID, LEVEL_ADMIN},//18
    {EP_AP_PASSWORD, LEVEL_ADMIN},//19
    {EP_AP_IP_VALUE, LEVEL_ADMIN},//20
    {EP_AP_MASK_VALUE, LEVEL_ADMIN},//21
    {EP_AP_GATEWAY_VALUE, LEVEL_ADMIN},//22
    {EP_AP_IP_MODE, LEVEL_ADMIN},//23
    {EP_AP_PHY_MODE, LEVEL_ADMIN},//24
    {EP_TARGET_FW, LEVEL_USER},//25
    {EP_TIMEZONE, LEVEL_USER},//26
    {EP_TIME_ISDST, LEVEL_USER},//27
    {EP_TIME_SERVER1, LEVEL_USER},//28
    {EP_TIME_SERVER2, LEVEL_USER},//29
    {EP_TIME_SERVER3, LEVEL_USER},//30
    {EP_IS_DIRECT_SD, LEVEL_USER},//31
    {EP_PRIMARY_SD, LEVEL_USER},//32
    {EP_SECONDARY_SD, LEVEL_USER},//33
    {EP_DIRECT_SD_CHECK, LEVEL_USER}, //34
    {EP_SD_CHECK_UPDATE_AT_BOOT, LEVEL_USER},//35
    {EP_OUTPUT_FLAG, LEVEL_USER},//36
    {EP_DHT_INTERVAL, LEVEL_USER},//37
    {EP_DHT_TYPE, LEVEL_USER},//38
    {EP_SD_SPEED_DIV, LEVEL_USER}//39
};
#define AUTH_ENTRY_NB 40

#define FLAG_BLOCK_M117 0x01
#define FLAG_BLOCK_OLED 0x02
#define FLAG_BLOCK_SERIAL 0x04
#define FLAG_BLOCK_WSOCKET 0x08
#define FLAG_BLOCK_TCP 0x010

//values
#define DEFAULT_MAX_WEB_PORT			65001
#define DEFAULT_MIN_WEB_PORT			1
#define DEFAULT_MAX_DATA_PORT			65001
#define DEFAULT_MIN_DATA_PORT			1

#define MAX_TRY 2000

//sizes
#define EEPROM_SIZE				1024 //max is 1024
#define MAX_SSID_LENGTH				32
#define MIN_SSID_LENGTH				1
#define MAX_PASSWORD_LENGTH 			64
//min size of password is 0 or upper than 8 char
//so let set min is 0
#define MIN_PASSWORD_LENGTH 			0
#define MAX_LOCAL_PASSWORD_LENGTH 			16
#define MIN_LOCAL_PASSWORD_LENGTH 			1
#define MAX_DATA_LENGTH				128
#define MIN_DATA_LENGTH				0
#define IP_LENGTH 				4
#define INTEGER_LENGTH 				4
#define MAX_HOSTNAME_LENGTH		32
#define MIN_HOSTNAME_LENGTH		1
#define WL_MAC_ADDR_LENGTH 6

#if defined(ASYNCWEBSERVER)
class AsyncResponseStream;
typedef  AsyncResponseStream ESPResponseStream; 
#else
class ESPResponseStream{
	public:
	bool header_sent;
	String buffer_web;
	ESPResponseStream(){header_sent=false;};
};
#endif

class CONFIG
{
public:
    static void wait (uint32_t milliseconds);
    static void wdtFeed();
    static byte output_flag;
#ifdef DHT_FEATURE
    static byte DHT_type;
    static int DHT_interval;
    static void InitDHT(bool refresh = false);
#endif
    static bool is_locked(byte flag);
    static bool is_direct_sd;
    static bool read_string (int pos, char byte_buffer[], int size_max);
    static bool read_string (int pos, String & sbuffer, int size_max);
    static bool read_buffer (int pos, byte byte_buffer[], int size_buffer);
    static bool read_byte (int pos, byte * value);
    static bool write_string (int pos, const char * byte_buffer);
    static bool write_string (int pos, const __FlashStringHelper *str);
    static bool write_buffer (int pos, const byte * byte_buffer, int size_buffer);
    static bool write_byte (int pos, const byte value);
    static bool reset_config();
    static void print_config (tpipe output, bool plaintext, ESPResponseStream  *espresponse = NULL);
    static bool SetFirmwareTarget (uint8_t fw);
    static void InitFirmwareTarget();
    static void InitOutput();
    static void InitDirectSD();
    static void InitPins();
    static bool InitBaudrate(long value = 0);
    static bool InitExternalPorts();
    static uint8_t GetFirmwareTarget();
    static const char* GetFirmwareTargetName();
    static const char* GetFirmwareTargetShortName();
    static bool isHostnameValid (const char * hostname);
    static bool isSSIDValid (const char * ssid);
    static bool isPasswordValid (const char * password);
    static bool isLocalPasswordValid (const char * password);
    static bool isIPValid (const char * IP);
    static char * intTostr (int value);
    static String formatBytes (uint32_t bytes);
    static char * mac2str (uint8_t mac [WL_MAC_ADDR_LENGTH]);
    static byte split_ip (const char * ptr, byte * part);
    static void esp_restart (bool async = false);
#if defined(TIMESTAMP_FEATURE)
    static void init_time_client();
#endif
private:
    static uint8_t FirmwareTarget;
};



#endif
